lms/services/grouping.py

```diff
@@ -0,0 +1,68 @@
+from lms.models import CanvasSection, Grouping
+from lms.models._hashed_id import hashed_id
+
+
+class GroupingService:
+    def __init__(self, db, application_instance_service, course_service):
+        self._db = db
+        self._application_instance = application_instance_service.get()
+        self._course_service = course_service
+
+    def upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
+                type=grouping.type,
+            )
+            .one_or_none()
+        )
+        if not db_grouping:
+            self._db.add(grouping)
+        else:
+            # Update any fields that might have changed
+            db_grouping.lms_name = grouping.lms_name
+            db_grouping.extra = grouping.extra
+
+        return db_grouping or grouping
+
+    def canvas_section(
+        self, tool_consumer_instance_guid, context_id, section_id, section_name
+    ):
```

Comment:
I think this should be called `upsert_canvas_section()` to clarify that it actually does an upsert.

Makes sense to take individual arguments here (rather than a single `models.CanvasSection` as the argument) since this is a convenience method that's doing the work of generating the `section_authority_provided_id` and `course_authority_provided_id` from other fields.